### PR TITLE
RFC: apimachinery: avoid modifying object when encoding TypeMeta as JSON

### PR DESCRIPTION
**Pull request by pohly**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

It's surprising that client-go Create calls don't treat the object as read-only. Encoding temporarily sets the kind of the object, which causes data races when the same object is used in parallel (for example, creating multiple different objects with a generated name).

A general solution is hard. However, a special solution for the most common case (API types which embed TypeMeta and JSON encoding) is possible: the underlying idea is that the type is asked to create a new TypeMeta instance, the kind is set there, then the encoder inlines the original struct and that separate TypeMeta in the same JSON object.

#### Which issue(s) this PR fixes:
Related-to: https://github.com/kubernetes/kubernetes/issues/82497

#### Special notes for your reviewer:

In practice, the implementation is a bit tricky because of the abstraction layers. Also, the JSON encoder doesn't inline interface types (https://github.com/golang/go/issues/6213#issuecomment-1500639389), so that part has to be done manually.

Further work is needed (covering more encoder options, more tests), but first I'd like to get feedback whether this approach is viable.

#### Does this PR introduce a user-facing change?

TBD

```release-note
```

### CODE REVIEW

File: staging/src/k8s.io/apimachinery/pkg/runtime/helper.go [hunk 1]
```diff
@@ -236,6 +236,19 @@ func (e WithVersionEncoder) Encode(obj Object, stream io.Writer) error {
 			gvk = preferredGVK
 		}
 	}
+	if gvk == oldGVK {
+		// gvk already set as intended, skip injecting it.
+		return e.Encoder.Encode(obj, stream)
```
Comment (pohly) [inline staging/src/k8s.io/apimachinery/pkg/runtime/helper.go:241]: Might be worth merging also without the other changes?

**Comment by pohly** [comment]
I also toyed with the idea of injecting the GVK into the object by implementing `MarshalJSON` for `TypeMeta` and then have that code check some global map to figure out whether it should return the type of the object or some override that gets set only for the duration of the `Encode` call.

That idea fell flat on its face because then all types inherit `MarshalJSON` as their own and only `TypeMeta` gets encoded without the object's fields.

File: staging/src/k8s.io/apimachinery/pkg/runtime/helper.go (hunk 1 above)
Comment (lavalamp) [reply staging/src/k8s.io/apimachinery/pkg/runtime/helper.go:241]: Yes please

File: staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go [hunk 2]
```diff
@@ -246,6 +248,63 @@ func (s *Serializer) doEncode(obj runtime.Object, w io.Writer) error {
...
+	// Write minus the trailing }.
+	len1 := buffer.Len()
+	if len1 < 2 || buffer.Bytes()[len1-2] != '}' || buffer.Bytes()[len1-1] != '\n' {
+		return errors.New("encoded struct did not have } at the end of the buffer")
+	}
+	if _, err := w.Write(buffer.Bytes()[:len1-2]); err != nil {
+		return err
+	}
+
+	// Encode kind struct and write it minus the leading {.
```
Comment (pohly) [inline staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go:285]:
```suggestion
	// Encode the second object and write it minus the leading {.
```

Comment (pohly) [reply staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go:285]: Kudos to @humblec who pointed this out to me.

**Comment by liggitt** [comment]
I thought WithVersionEncoder was used for all output types, not just json

**Comment by pohly** [comment]
This might become easier with ["encoding/json/v2"](https://github.com/golang/go/discussions/63397) - worth checking, I'm having a look.

**Comment by pohly** [comment]
Yes, "Caller-specified customization" should fix our problem. There's a slight problem with abstractions (the JSON marshaler needs to be set for a specific type, but the apimachinery doesn't know that type), but that should be solvable.

**Comment by pohly** [comment]
I'll come back to this once we can use json/v2 and "Caller-specified customization". In the meantime I changed the code so that `WithVersionEncoder` and `TypeMeta` can collaborate on injecting the desired GVK. As a proof-of-concept for the new interface, this "side loading" is done for protobuf by modifying the `generated.pb.go` so that it gets wrapped by custom code which supports overriding the `TypeMeta` instance during encoding.

Still pretty hacky, but perhaps it can be come cleaner...

File: staging/src/k8s.io/apimachinery/pkg/runtime/helper.go (hunk 1 above)
Comment (pohly) [reply staging/src/k8s.io/apimachinery/pkg/runtime/helper.go:241]: => https://github.com/kubernetes/kubernetes/pull/121771

**Pull request closed without merging**